hathor/event/websocket/factory.py

```diff
@@ -0,0 +1,123 @@
+from pydantic import ValidationError
+from structlog import get_logger
+
+from hathor.event import BaseEvent
+from hathor.event.storage import EventStorage
+from hathor.event.websocket.protocol import EventWebsocketProtocol
+from hathor.event.websocket.request import StreamRequest
+from hathor.event.websocket.response import BadRequestResponse, EventResponse, EventWebSocketNotRunningResponse
+from hathor.util import json_dumpb
+
+logger = get_logger()
+
+
+class EventWebsocketFactory(WebSocketServerFactory):
+    """ Websocket that will handle events
+    """
+
+    protocol = EventWebsocketProtocol
+    _is_running = False
+    _latest_event_id: Optional[int] = None
+
+    def __init__(self, event_storage: EventStorage):
+        super().__init__()
+        self.log = logger.new()
+        self._event_storage = event_storage
+        self._connections: Set[EventWebsocketProtocol] = set()
+
+        latest_event = self._event_storage.get_last_event()
+
+        if latest_event is not None:
+            self._latest_event_id = latest_event.id
+
+    def start(self):
+        """Start the WebSocket server. Required to be able to send events."""
+        self._is_running = True
+
+    def stop(self):
+        """Stop the WebSocket server. No events can be sent."""
+        self._is_running = False
+
+        for connection in self._connections:
+            connection.sendClose()
+
+        self._connections.clear()
+
+    def broadcast_event(self, event: BaseEvent) -> None:
+        """Broadcast the event to each registered client."""
+        self._latest_event_id = event.id
+
+        for connection in self._connections:
+            if event.id == connection.next_event_id:
+                self._send_event_to_connection(connection, event)
+
+    def register(self, connection: EventWebsocketProtocol) -> None:
+        """Registers a client. Called when a ws connection is opened (after handshaking)."""
+        if not self._is_running:
+            response = EventWebSocketNotRunningResponse().dict()
+            payload = json_dumpb(response)
+
+            return connection.sendMessage(payload)
+
+        self.log.info('registering connection', client_peer=connection.client_peer)
+
+        self._connections.add(connection)
+
+    def unregister(self, connection: EventWebsocketProtocol) -> None:
+        """Unregisters a client. Called when a ws connection is closed."""
+        self.log.info('unregistering connection', client_peer=connection.client_peer)
+        self._connections.discard(connection)
+
+    def handle_valid_request(self, connection: EventWebsocketProtocol, request: StreamRequest) -> None:
+        """Handle a valid client request."""
+        connection.last_received_event_id = request.last_received_event_id
+        connection.available_window_size += request.window_size_increment
+
+        events = self._event_storage.iter_from_event(connection.next_event_id)
+
+        for event in events:
```

Comment:
I feel this loop might lock the event loop for a long time (and we can't do it). Besides that, the current implementation is different from what I understood. The client would not open many event streamings. In fact, the client would open the stream and would send many ACK to keep the flow going.

There are different ways of implement this, and the major implementation details for me are the following:

1) Send one event per call, letting the event loop run free.
2) Keep sending events until the window is zeroed or the next expected id is not available.
3) When we receive an ACK, we must try to send the next event again because the available window might have changed.

If the client wants to pause the streaming, it just have to send an ACK message with window=0.

Reply:
Got it. Actually, this was a doubt I had about how Twisted works. I had understood that `sendMessage()` would just schedule the send, but not block the event loop.

I've changed the code to be more similar to what you described, using Twisted's `callLater()` method. Could you please check it again?

Reply:
If you check the source-code of `sendMessage(...)` [1], you will see that it does many things before calling `sendFrame(...)`, which ends up calling `transport.write(...)`, which sends the byte sequence to the transport's buffer. It seems there's a lot happening for each call. Anyway, regardless of the `sendMessage(...)` performance, your method is not async and it will only return to the event loop after your loop ends.

I read the code it's still different from what I was expecting. I'll add more details on a separate message.

[1] https://autobahn.readthedocs.io/en/latest/_modules/autobahn/websocket/protocol.html#WebSocketProtocol.sendMessage

Reply:
In my view, your implement is the same as sending batches of events, and it's different from what I proposed. In my proposal, client's ACK would be different from last expected id. In your code, every ACK will trigger a new batch of events, which would stop a streaming of events from occurring.

Note that there's latency and packet loss between the client and server; when packets are lost and TCP will need some time to retransmit them. For clarity, let's get an extreme case: a high latency with high throughput connection. For example, 2,000 ms of latency with 50Mbps of throughput. Let's also assume that the client has enough computing power to process a long window of events (e.g., 100k events) in real time. In this case, the server is able to send hundreds (if not thousands) of events before receiving the next ACK.

In my view, ACK and Start Streaming are very different messages sent by the client. The Start Streaming would set the starting point. Note that the use case would persist the latest received event and would use it in case of a restart. The ACK message would be used only for flow control. If we hadn't the need for flow control, we could completely remove the ACK message (but that's not the case).

We can discuss it on a meeting tomorrow. Maybe after the daily?

Reply:
Done!

Reply:
It has already been re-implemented. Thanks, @glevco .